README.md
# lsconntrack

[![Build Status](https://travis-ci.org/yuuki/lsconntrack.png?branch=master)][travis]
[![Go Report Card](https://goreportcard.com/badge/github.com/yuuki/lsconntrack)][goreportcard]

[license]: https://github.com/yuuki/lsconntrack/blob/master/LICENSE
[travis]: https://travis-ci.org/yuuki/lsconntrack
[goreportcard]: (https://goreportcard.com/report/github.com/yuuki/lsconntrack)

lsconntrack prints `host flows` (aggregated connection flows to the same source or destination ports) tracked by Linux netfilter conntrack and enables you to simply grasp the network relationship between localhost and other hosts.

friend: [yuuki/lstf](https://github.com/yuuki/lstf)

## Features

- Distinction of `active opens` and `passive opens`
- Print also packets and bytes of each flows (the absolute values are meaningless)
- Go portability
- Filter by ports (--active-ports and --passive-ports)
- stdin support (combination with [conntrack-tools](http://conntrack-tools.netfilter.org/))
- JSON support
- TCP support only
- TODO: streaming support

## Environment

- Linux only
  - Tested on CentOS5, Debian7, Debian8

## Setup

- Load ip_conntrack / nf_conntrack module
- Download the binary from https://github.com/yuuki/lsconntrack/releases .

## How to use

```shell
$ lsconntrack -n
Local Address:Port   <-->   Peer Address:Port     Inpkts  Inbytes   Outpkts Outbytes
localhost:many       -->    10.0.1.10:3306        5521792 123258667 5423865 282041045
localhost:many       -->    10.0.1.11:3306        58800   3062451   58813   3061627
localhost:many       -->    10.0.1.20:8080        123     169638    62      3580
localhost:80         <--    10.0.2.10:80          23      6416      25      25460
localhost:80         <--    10.0.2.11:80          38      8574      34      32752
```

```shell
# Prints active open connections from localhost to destination hosts.
$ lsconntrack --active
Local Address:Port   <-->   Peer Address:Port     Inpkts  Inbytes   Outpkts Outbytes
localhost:many       -->    10.0.1.10:3306        5521792 123258667 5423865 282041045
localhost:many       -->    10.0.1.11:3306        58800   3062451   58813   3061627
localhost:many       -->    10.0.1.20:8080        123     169638    62      3580
...
```

```shell
# Prints passive open connections from destination hosts to localhost.
$ lsconntrack --passive
Local Address:Port   <-->   Peer Address:Port   Inpkts  Inbytes   Outpkts Outbytes
localhost:80         <--    10.0.2.10:many      23      6416      25      25460
localhost:80         <--    10.0.2.11:many      38      8574      34      32752
...
```

### filter by port

```shell
$ lsconntrack --active --aport 3306 --aport 11211
```

### via stdin

```shell
$ cat /proc/net/nf_conntrack | lsconntrack --stdin
```

### JSON format

```shell
$ lsconntrack --json | jq -r -M '.'
[
  {
    "direction": "active",
    "local": {
      "Addr": "localhost",
      "Port": "many"
    },
    "peer": {
      "addr": "10.0.100.1",
      "port": "3306"
    },
    "stat": {
      "total_inbound_packets": 1491,
      "total_inbound_bytes": 1480239,
      "total_outbound_packets": 1537,
      "total_outbound_bytes": 520613
    }
  },
  {
    "direction": "passive",
    "local": {
      "addr": "localhost",
      "port": "80"
    },
    "peer": {
      "addr": "10.0.200.1",
      "port": "many"
    },
    "stat": {
      "total_inbound_packets": 1491,
      "total_inbound_bytes": 1480239,
      "total_outbound_packets": 1537,
      "total_outbound_bytes": 520613
    }
  },
  ...
]
```

## License

[MIT][license]

## Author

[yuuki](https://github.com/y_uuki)


CHANGELOG.md
# Changelog

## [v0.3.0](https://github.com/yuuki/lsconntrack/compare/v0.2.0...v0.3.0) (2018-03-04)

* Change json format [#10](https://github.com/yuuki/lsconntrack/pull/10) ([yuuki](https://github.com/yuuki))
* Improve JSON output [#9](https://github.com/yuuki/lsconntrack/pull/9) ([yuuki](https://github.com/yuuki))

## [v0.2.0](https://github.com/yuuki/lsconntrack/compare/v0.1.2...v0.2.0) (2018-03-04)

* Add --numeric option [#8](https://github.com/yuuki/lsconntrack/pull/8) ([yuuki](https://github.com/yuuki))
* Print active and passive host flow if both --active and --pasive is not specified [#7](https://github.com/yuuki/lsconntrack/pull/7) ([yuuki](https://github.com/yuuki))
* Simplify mode distinction to remove entries [#6](https://github.com/yuuki/lsconntrack/pull/6) ([yuuki](https://github.com/yuuki))
* Print as flow arrow [#5](https://github.com/yuuki/lsconntrack/pull/5) ([yuuki](https://github.com/yuuki))

## [v0.1.2](https://github.com/yuuki/lsconntrack/compare/v0.1.1...v0.1.2) (2018-03-02)

* Enable to work --active without filtering by ports arguments [#4](https://github.com/yuuki/lsconntrack/pull/4) ([yuuki](https://github.com/yuuki))

## [v0.1.1](https://github.com/yuuki/lsconntrack/compare/v0.1.0...v0.1.1) (2018-03-01)

* Fix panic that occurs when no pakcets or bytes item [#3](https://github.com/yuuki/lsconntrack/pull/3) ([yuuki](https://github.com/yuuki))
* Adopt listening ports as default ports [#2](https://github.com/yuuki/lsconntrack/pull/2) ([yuuki](https://github.com/yuuki))

## [v0.1.0](https://github.com/yuuki/lsconntrack/compare/...v0.1.0) (2018-02-27)

* Initial release


cli_test.go
package main

import (
	"bytes"
	"strings"
	"testing"
)

func TestRun_global(t *testing.T) {
	tests := []struct {
		desc           string
		arg            string
		expectedStatus int
		expectedSubOut string
		expectedSubErr string
	}{
		{
			desc:           "undefined flag",
			arg:            "lsconntrack --undefined",
			expectedStatus: exitCodeFlagParseError,
			expectedSubErr: "flag provided but not defined",
		},
	}
	for _, tc := range tests {
		outStream, errStream := new(bytes.Buffer), new(bytes.Buffer)
		cli := &CLI{outStream: outStream, errStream: errStream}
		args := strings.Split(tc.arg, " ")

		status := cli.Run(args)
		if status != tc.expectedStatus {
			t.Errorf("desc: %q, status should be %v, not %v", tc.desc, tc.expectedStatus, status)
		}

		if !strings.Contains(outStream.String(), tc.expectedSubOut) {
			t.Errorf("desc: %q, subout should contain %q, got %q", tc.desc, tc.expectedSubOut, outStream.String())
		}
		if !strings.Contains(errStream.String(), tc.expectedSubErr) {
			t.Errorf("desc: %q, subout should contain %q, got %q", tc.desc, tc.expectedSubErr, errStream.String())
		}
	}
}


conntrack/util_test.go
package conntrack

import "testing"

func TestContains(t *testing.T) {
	tests := []struct {
		in  []string
		s   string
		out bool
	}{
		{[]string{"3306", "11211", "80", "443", "6379"}, "11211", true},
		{[]string{"3306", "11211", "80", "443", "6379"}, "9000", false},
		{[]string{}, "3306", false},
	}

	for _, tt := range tests {
		if contains(tt.in, tt.s) != tt.out {
			t.Errorf("contains(%v, %v) == %v", tt.in, tt.s, tt.out)
		}
	}
}


conntrack/conntrack_test.go
package conntrack

import "testing"

func TestParseLine(t *testing.T) {
	t.Run("[UNREPLIRED]", func(t *testing.T) {
		line := "tcp      6 367755 ESTABLISHED src=10.0.0.1 dst=10.0.0.2 sport=3306 dport=38205 packets=1 bytes=52 [UNREPLIED] src=10.0.0.2 dst=10.0.0.1 sport=38205 dport=3306 packets=0 bytes=0 mark=0 secmark=0 use=1"
		rstat := parseLine(line)
		if rstat.originalSaddr != "10.0.0.1" {
			t.Errorf("OriginalSaddr should be 10.0.0.1, not %v", rstat.originalSaddr)
		}
		if rstat.originalDaddr != "10.0.0.2" {
			t.Errorf("OriginalDaddr should be 10.0.0.2, not %v", rstat.originalDaddr)
		}
		if rstat.originalSport != "3306" {
			t.Errorf("OriginalSport should be 3306, not %v", rstat.originalSport)
		}
		if rstat.originalDport != "38205" {
			t.Errorf("OriginalSport should be 38205, not %v", rstat.originalDport)
		}
		if rstat.originalPackets != 1 {
			t.Errorf("OriginalPackets should be 1, not %v", rstat.originalPackets)
		}
		if rstat.originalBytes != 52 {
			t.Errorf("OriginalBytes should be 52, not %v", rstat.originalBytes)
		}
		if rstat.replySaddr != "10.0.0.2" {
			t.Errorf("ReplySaddr should be 10.0.0.2, not %v", rstat.replySaddr)
		}
		if rstat.replyDaddr != "10.0.0.1" {
			t.Errorf("ReplySaddr should be 10.0.0.1, not %v", rstat.replyDaddr)
		}
		if rstat.replySport != "38205" {
			t.Errorf("ReplySaddr should be 38205, not %v", rstat.replySport)
		}
		if rstat.replyDport != "3306" {
			t.Errorf("ReplySaddr should be 3306, not %v", rstat.replyDport)
		}
		if rstat.replyPackets != 0 {
			t.Errorf("ReplyPackets should be 0, not %v", rstat.replyPackets)
		}
		if rstat.replyBytes != 0 {
			t.Errorf("ReplyBytes should be 0, not %v", rstat.replyBytes)
		}
	})

	t.Run("[ASSURED]", func(t *testing.T) {
		line := "tcp      6 5 CLOSE src=10.0.0.10 dst=10.0.0.11 sport=41143 dport=443 packets=3 bytes=164 src=10.0.0.11 dst=10.0.0.10 sport=443 dport=41143 packets=1 bytes=60 [ASSURED] mark=0 secmark=0 use=1"
		rstat := parseLine(line)
		if rstat.originalSaddr != "10.0.0.10" {
			t.Errorf("OriginalSaddr should be 10.0.0.10, not %v", rstat.originalSaddr)
		}
		if rstat.originalDaddr != "10.0.0.11" {
			t.Errorf("OriginalDaddr should be 10.0.0.11, not %v", rstat.originalDaddr)
		}
		if rstat.originalSport != "41143" {
			t.Errorf("OriginalSport should be 41143, not %v", rstat.originalSport)
		}
		if rstat.originalDport != "443" {
			t.Errorf("OriginalSport should be 443, not %v", rstat.originalDport)
		}
		if rstat.originalPackets != 3 {
			t.Errorf("OriginalPackets should be 3, not %v", rstat.originalPackets)
		}
		if rstat.originalBytes != 164 {
			t.Errorf("OriginalBytes should be 164, not %v", rstat.originalBytes)
		}
		if rstat.replySaddr != "10.0.0.11" {
			t.Errorf("ReplySaddr should be 10.0.0.11, not %v", rstat.replySaddr)
		}
		if rstat.replyDaddr != "10.0.0.10" {
			t.Errorf("ReplySaddr should be 10.0.0.10, not %v", rstat.replyDaddr)
		}
		if rstat.replySport != "443" {
			t.Errorf("ReplySaddr should be 443, not %v", rstat.replySport)
		}
		if rstat.replyDport != "41143" {
			t.Errorf("ReplySaddr should be 41143, not %v", rstat.replyDport)
		}
		if rstat.replyPackets != 1 {
			t.Errorf("ReplyPackets should be 1, not %v", rstat.replyPackets)
		}
		if rstat.replyBytes != 60 {
			t.Errorf("ReplyBytes should be 60, not %v", rstat.replyBytes)
		}
	})
}


cli.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strconv"
	"text/tabwriter"

	"github.com/yuuki/lsconntrack/conntrack"
	"github.com/yuuki/lsconntrack/netutil"
)

const (
	exitCodeOK             = 0
	exitCodeFlagParseError = 10 + iota
	exitCodeArgumentsError
	exitCodeParseConntrackError
	exitCodePrintError
	exitCodeUnreachableError
)

type portslice []string

func (s *portslice) String() string {
	return fmt.Sprintf("%s", *s)
}

func (s *portslice) Set(value string) error {
	if _, err := strconv.Atoi(value); err != nil {
		return fmt.Errorf("%s is not number", value)
	}
	*s = append(*s, value)
	return nil
}

// CLI is the command line object.
type CLI struct {
	// outStream and errStream are the stdout and stderr
	// to write message from the CLI.
	outStream, errStream io.Writer
}

// Run execute the main process.
// It returns exit code.
func (c *CLI) Run(args []string) int {
	log.SetOutput(c.errStream)

	var (
		active, passive           bool
		activePorts, passivePorts portslice
		numeric                   bool
		stdin                     bool
		json                      bool
		ver                       bool
	)
	flags := flag.NewFlagSet("lsconntrack", flag.ContinueOnError)
	flags.SetOutput(c.errStream)
	flags.Usage = func() {
		fmt.Fprint(c.errStream, helpText)
	}
	flags.BoolVar(&active, "a", false, "")
	flags.BoolVar(&active, "active", false, "")
	flags.BoolVar(&passive, "p", false, "")
	flags.BoolVar(&passive, "passive", false, "")
	flags.Var(&activePorts, "aport", "")
	flags.Var(&activePorts, "active-port", "")
	flags.Var(&passivePorts, "pport", "")
	flags.Var(&passivePorts, "passive-port", "")
	flags.BoolVar(&numeric, "n", false, "")
	flags.BoolVar(&numeric, "numeric", false, "")
	flags.BoolVar(&stdin, "stdin", false, "")
	flags.BoolVar(&json, "json", false, "")
	flags.BoolVar(&ver, "version", false, "")
	if err := flags.Parse(args[1:]); err != nil {
		return exitCodeFlagParseError
	}

	if ver {
		fmt.Fprintf(c.errStream, "%s version %s, build %s, date %s \n", name, version, commit, date)
		return exitCodeOK
	}

	var mode conntrack.FlowDirection
	if active {
		mode |= conntrack.FlowActive
	}
	if passive {
		mode |= conntrack.FlowPassive
	}
	if !active && !passive {
		mode = conntrack.FlowActive | conntrack.FlowPassive
	}

	var r io.Reader
	if stdin {
		r = os.Stdin
	} else {
		path := netutil.FindConntrackPath()
		if path == "" {
			log.Println("not found conntrack entries path: Please load conntrack module")
			return exitCodeParseConntrackError
		}
		f, err := os.Open(path)
		if err != nil {
			log.Printf("failed to open %v: %v\n", path, err)
			return exitCodeParseConntrackError
		}
		defer f.Close()
		r = f
	}

	if mode&conntrack.FlowPassive != 0 && len(passivePorts) == 0 {
		var err error
		passivePorts, err = netutil.LocalListeningPorts()
		if err != nil {
			log.Printf("failed to get local listening ports: %v\n", err)
			return exitCodeParseConntrackError
		}
	}

	flows, err := conntrack.ParseEntries(r, conntrack.FilterPorts{
		Active:  activePorts,
		Passive: passivePorts,
	})
	if err != nil {
		log.Println(err)
		return exitCodeParseConntrackError
	}
	if json {
		if err := c.PrintHostFlowsAsJSON(flows, numeric, mode); err != nil {
			log.Println(err)
			return exitCodePrintError
		}
	} else {
		c.PrintHostFlows(flows, numeric, mode)
	}

	return exitCodeOK
}

// PrintHostFlows prints the host flows.
func (c *CLI) PrintHostFlows(flows conntrack.HostFlows, numeric bool, direction conntrack.FlowDirection) {
	// Format in tab-separated columns with a tab stop of 8.
	tw := tabwriter.NewWriter(c.outStream, 0, 8, 0, '\t', 0)
	fmt.Fprintln(tw, "Local Address:Port\t <--> \tPeer Address:Port \tInpkts \tInbytes \tOutpkts \tOutbytes")
	for _, flow := range flows {
		if flow.HasDirection(direction) {
			continue
		}
		if !numeric {
			flow.ReplaceLookupedName()
		}
		fmt.Fprintln(tw, flow)
	}
	tw.Flush()
}

// PrintHostFlowsAsJSON prints the host flows as json format.
func (c *CLI) PrintHostFlowsAsJSON(flows conntrack.HostFlows, numeric bool, direction conntrack.FlowDirection) error {
	for key, flow := range flows {
		if !numeric {
			flow.ReplaceLookupedName()
		}
		if flow.HasDirection(direction) {
			delete(flows, key)
		}
	}
	b, err := json.Marshal(flows)
	if err != nil {
		return err
	}
	c.outStream.Write(b)
	return nil
}

var helpText = `Usage: lsconntrack [options]

  Print host flows between localhost and other hosts

Options:
  --active, -a              print active-open host flows (from localhost to other host).
  --passive, -p             print passive-open host flows (from other host to localhost).
  --active-port, --aport    output filter by active-open destination ports
  --passive-port, --pport   output filter by localhost listening ports (default: all listening local ports)
  --numeric, -n             show numerical addresses instead of trying to determine symbolic host, port names.
  --stdin                   input conntrack entries via stdin
  --json                    print results as json format
  --version, -v	            print version
  --help, -h                print help
`


main.go
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"text/tabwriter"

	"github.com/yuuki/lsconntrack/conntrack"
)

const (
	exitCodeOK = iota
	exitCodeFlagParseError
	exitCodeArgumentsError
	exitCodeParseConntrackError
)

// Run execute the main process.
// It returns exit code.
func Run(args []string) int {
	var (
		activeMode  bool
		passiveMode bool
		stdin       bool
	)
	flags := flag.NewFlagSet("lsconntrack", flag.ContinueOnError)
	flags.SetOutput(os.Stderr)
	flags.Usage = func() {
		fmt.Fprint(os.Stderr, helpText)
	}
	flags.BoolVar(&activeMode, "a", false, "")
	flags.BoolVar(&activeMode, "active", false, "")
	flags.BoolVar(&passiveMode, "p", false, "")
	flags.BoolVar(&passiveMode, "passive", false, "")
	flags.BoolVar(&stdin, "stdin", false, "")
	if err := flags.Parse(args[1:]); err != nil {
		return exitCodeFlagParseError
	}

	ports := flags.Args()
	if len(ports) == 0 {
		log.Println("ports required")
		fmt.Fprint(os.Stderr, helpText)
		return exitCodeArgumentsError
	}

	var r io.Reader
	if stdin {
		r = os.Stdin
	} else {
		path := conntrack.FindEntryPath()
		if path == "" {
			log.Println("not found conntrack entries path: Please load conntrack module")
			return exitCodeParseConntrackError
		}
		f, err := os.Open(path)
		if err != nil {
			log.Printf("failed to open %v: %v\n", path, err)
			return exitCodeParseConntrackError
		}
		defer f.Close()
		r = f
	}

	var connStat conntrack.ConnStatByAddrPort
	if activeMode {
		var err error
		connStat, err = conntrack.ParseEntries(r, conntrack.ConnActive, ports)
		if err != nil {
			log.Println(err)
			return exitCodeParseConntrackError
		}
	} else if passiveMode {
		var err error
		connStat, err = conntrack.ParseEntries(r, conntrack.ConnPassive, ports)
		if err != nil {
			log.Println(err)
			return exitCodeParseConntrackError
		}
	}

	// Format in tab-separated columns with a tab stop of 8.
	tw := tabwriter.NewWriter(os.Stdout, 0, 8, 0, '\t', 0)
	for _, stat := range connStat {
		hostnames, _ := net.LookupAddr(stat.Addr)
		var hostname string
		if len(hostnames) > 0 {
			hostname = hostnames[0]
		}
		fmt.Fprintf(tw, "%s:%s\t%s\t%d\t%d\t%d\t%d\n", stat.Addr, stat.Port, hostname, stat.TotalInboundPackets, stat.TotalInboundBytes, stat.TotalOutboundPackets, stat.TotalOutboundBytes)
	}
	tw.Flush()
	return exitCodeOK
}

func main() {
	os.Exit(Run(os.Args))
}

var helpText = `Usage: lsconntrack [options] port...

  Print aggregated connections between localhost and other hosts

Options:
  --active, -a      print aggregated connections localhost to destination
  --passive, -p     print aggregated connections source to localhost
  --numeric, -n     show numerical addresses instead of trying to determine symbolic host, port names.
  --stdin           input conntrack entries via stdin
  --version, -v		print version
  --help, -h        print help
`


conntrack/conntrack.go
package conntrack

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"strconv"
	"strings"

	"github.com/yuuki/lsconntrack/netutil"
)

// FlowDirection are bitmask that represents both Active or Passive.
type FlowDirection int

const (
	// FlowUnknown are unknown flow.
	FlowUnknown FlowDirection = 1 << iota
	// FlowActive are 'active open'.
	FlowActive
	// FlowPassive are 'passive open'
	FlowPassive
)

// MarshalJSON returns human readable `mode` format.
func (c FlowDirection) MarshalJSON() ([]byte, error) {
	switch c {
	case FlowActive:
		return json.Marshal("active")
	case FlowPassive:
		return json.Marshal("passive")
	case FlowUnknown:
		return json.Marshal("unknown")
	}
	return nil, errors.New("unreachable code")
}

// FilterPorts are ports to filter output.
type FilterPorts struct {
	Active  []string
	Passive []string
}

// flow represents statistics of a connection to other host and port.
type flow struct {
	originalSaddr   string
	originalDaddr   string
	originalSport   string
	originalDport   string
	originalPackets int64
	originalBytes   int64
	replySaddr      string
	replyDaddr      string
	replySport      string
	replyDport      string
	replyPackets    int64
	replyBytes      int64
}

// HostFlowStat represents statistics of a host flow.
type HostFlowStat struct {
	TotalInboundPackets  int64 `json:"total_inbound_packets"`
	TotalInboundBytes    int64 `json:"total_inbound_bytes"`
	TotalOutboundPackets int64 `json:"total_outbound_packets"`
	TotalOutboundBytes   int64 `json:"total_outbound_bytes"`
}

// String returns the string representation of the HostFlowStat.
func (s *HostFlowStat) String() string {
	return fmt.Sprintf("%d \t%d \t%d \t%d", s.TotalInboundPackets, s.TotalInboundBytes, s.TotalOutboundPackets, s.TotalOutboundBytes)
}

// AddrPort are <addr>:<port>
type AddrPort struct {
	Addr string `json:"addr"`
	Port string `json:"port"`
}

// String returns the string representation of the AddrPort.
func (a *AddrPort) String() string {
	return net.JoinHostPort(a.Addr, a.Port)
}

// HostFlow represents a `host flow`.
type HostFlow struct {
	Direction FlowDirection `json:"direction"`
	Local     *AddrPort     `json:"local"`
	Peer      *AddrPort     `json:"peer"`
	Stat      *HostFlowStat `json:"stat"`
}

// HasDirection returns whether .
func (f *HostFlow) HasDirection(dire FlowDirection) bool {
	return f.Direction&dire == 0
}

// String returns the string representation of HostFlow.
func (f *HostFlow) String() string {
	switch f.Direction {
	case FlowActive:
		return fmt.Sprintf("%s\t --> \t%s \t%s", f.Local, f.Peer, f.Stat)
	case FlowPassive:
		return fmt.Sprintf("%s\t <-- \t%s \t%s", f.Local, f.Peer, f.Stat)
	}
	return ""
}

// ReplaceLookupedName replaces f.Addr into lookuped name.
func (f *HostFlow) ReplaceLookupedName() {
	f.Peer.Addr = netutil.ResolveAddr(f.Peer.Addr)
}

// UniqKey returns the unique key for connections aggregation
func (f *HostFlow) UniqKey() string {
	return fmt.Sprintf("%d-%s-%s", f.Direction, f.Local, f.Peer)
}

// HostFlows represents a group of host flow by unique key.
type HostFlows map[string]*HostFlow

func (hf HostFlows) insert(flow *HostFlow) {
	key := flow.UniqKey()
	if _, ok := hf[key]; !ok {
		hf[key] = flow
		return
	}
	switch flow.Direction {
	case FlowActive:
		hf[key].Stat.TotalInboundPackets += flow.Stat.TotalInboundPackets
		hf[key].Stat.TotalInboundBytes += flow.Stat.TotalInboundBytes
		hf[key].Stat.TotalOutboundPackets += flow.Stat.TotalOutboundPackets
		hf[key].Stat.TotalOutboundBytes += flow.Stat.TotalOutboundBytes
	case FlowPassive:
		hf[key].Stat.TotalInboundPackets += flow.Stat.TotalInboundPackets
		hf[key].Stat.TotalInboundBytes += flow.Stat.TotalInboundBytes
		hf[key].Stat.TotalOutboundPackets += flow.Stat.TotalOutboundPackets
		hf[key].Stat.TotalOutboundBytes += flow.Stat.TotalOutboundBytes
	}
	return
}

// MarshalJSON returns list formats not map.
func (hf HostFlows) MarshalJSON() ([]byte, error) {
	list := make([]HostFlow, 0, len(hf))
	for _, f := range hf {
		list = append(list, *f)
	}
	return json.Marshal(list)
}

// toHostFlow converts into HostFlow.
func (f *flow) toHostFlow(localAddrs []string, fports FilterPorts) *HostFlow {
	var (
		direction  FlowDirection
		addr, port string
	)
	for _, localAddr := range localAddrs {
		// not filter by ports on ActiveOpen connection if ports is empty
		if f.originalSaddr == localAddr && (len(fports.Active) == 0 || contains(fports.Active, f.originalDport)) {
			direction, addr, port = FlowActive, f.originalDaddr, f.originalDport
			break
		}
		if f.replyDaddr == localAddr && (len(fports.Active) == 0 || contains(fports.Active, f.replySport)) {
			direction, addr, port = FlowActive, f.replySaddr, f.replySport
			break
		}
		if f.originalDaddr == localAddr && contains(fports.Passive, f.originalDport) {
			direction, addr, port = FlowPassive, f.originalSaddr, f.originalDport // not OriginalSport
			break
		}
		if f.replySaddr == localAddr && contains(fports.Passive, f.replySport) {
			direction, addr, port = FlowPassive, f.replyDaddr, f.replySport // not ReplyDport
			break
		}
	}
	switch direction {
	case FlowUnknown:
		return nil
	case FlowActive:
		return &HostFlow{
			Direction: FlowActive,
			Local:     &AddrPort{Addr: "localhost", Port: "many"},
			Peer:      &AddrPort{Addr: addr, Port: port},
			Stat: &HostFlowStat{
				TotalInboundPackets:  f.replyPackets,
				TotalInboundBytes:    f.replyBytes,
				TotalOutboundPackets: f.originalPackets,
				TotalOutboundBytes:   f.originalBytes,
			},
		}
	case FlowPassive:
		return &HostFlow{
			Direction: FlowPassive,
			Local:     &AddrPort{Addr: "localhost", Port: port},
			Peer:      &AddrPort{Addr: addr, Port: "many"},
			Stat: &HostFlowStat{
				TotalInboundPackets:  f.originalPackets,
				TotalInboundBytes:    f.originalBytes,
				TotalOutboundPackets: f.replyPackets,
				TotalOutboundBytes:   f.replyBytes,
			},
		}
	}
	return nil
}

func parseLine(line string) *flow {
	flow := &flow{}
	fields := strings.Fields(line)
	if len(fields) == 0 {
		log.Fatalf("unexpected line: %s\n", line)
	}
	if fields[0] != "tcp" {
		return nil
	}
	var packets, bytes bool
	if strings.Contains(line, "packets=") {
		packets = true
	}
	if strings.Contains(line, "bytes=") {
		bytes = true
	}
	if strings.Contains(line, "[UNREPLIED]") {
		// tcp      6 367755 ESTABLISHED src=10.0.0.1 dst=10.0.0.2 sport=3306 dport=38205 packets=1 bytes=52 [UNREPLIED] src=10.0.0.2 dst=10.0.0.1 sport=38205 dport=3306 packets=0 bytes=0 mark=0 secmark=0 use=1
		i := 4
		flow.originalSaddr = strings.Split(fields[i], "=")[1]
		flow.originalDaddr = strings.Split(fields[i+1], "=")[1]
		flow.originalSport = strings.Split(fields[i+2], "=")[1]
		flow.originalDport = strings.Split(fields[i+3], "=")[1]
		i = i + 4
		if bytes {
			flow.originalPackets, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		if packets {
			flow.originalBytes, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		i = i + 1
		flow.replySaddr = strings.Split(fields[i], "=")[1]
		flow.replyDaddr = strings.Split(fields[i+1], "=")[1]
		flow.replySport = strings.Split(fields[i+2], "=")[1]
		flow.replyDport = strings.Split(fields[i+3], "=")[1]
		i = i + 4
		if packets {
			flow.replyPackets, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		if bytes {
			flow.replyBytes, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		return flow
	} else if strings.Contains(line, "[ASSURED]") {
		// tcp      6 5 CLOSE src=10.0.0.10 dst=10.0.0.11 sport=41143 dport=443 packets=3 bytes=164 src=10.0.0.11 dst=10.0.0.10 sport=443 dport=41143 packets=1 bytes=60 [ASSURED] mark=0 secmark=0 use=1
		i := 4
		flow.originalSaddr = strings.Split(fields[i], "=")[1]
		flow.originalDaddr = strings.Split(fields[i+1], "=")[1]
		flow.originalSport = strings.Split(fields[i+2], "=")[1]
		flow.originalDport = strings.Split(fields[i+3], "=")[1]
		i = i + 4
		if packets {
			flow.originalPackets, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		if bytes {
			flow.originalBytes, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		flow.replySaddr = strings.Split(fields[i], "=")[1]
		flow.replyDaddr = strings.Split(fields[i+1], "=")[1]
		flow.replySport = strings.Split(fields[i+2], "=")[1]
		flow.replyDport = strings.Split(fields[i+3], "=")[1]
		i = i + 4
		if packets {
			flow.replyPackets, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		if bytes {
			flow.replyBytes, _ = strconv.ParseInt(strings.Split(fields[i], "=")[1], 10, 64)
			i++
		}
		return flow
	}
	return nil
}

// ParseEntries parses '/proc/net/nf_conntrack or /proc/net/ip_conntrack'.
func ParseEntries(r io.Reader, fports FilterPorts) (HostFlows, error) {
	localAddrs, err := netutil.LocalIPAddrs()
	if err != nil {
		return nil, err
	}
	hostFlows := HostFlows{}
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := scanner.Text()
		flow := parseLine(line)
		if flow == nil {
			continue
		}
		hostFlow := flow.toHostFlow(localAddrs, fports)
		if hostFlow == nil {
			continue
		}
		hostFlows.insert(hostFlow)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return hostFlows, nil
}
